`runTest` doesn't report test failure if a non-cancellable coroutine happens to be hanging

I'm using `kotlinx-coroutines-test` 1.6.0-RC.

If some non-cancellable coroutine code is hanging, exceptions or assertion errors from the test are never reported despite being reached:

```kotlin
@Test
fun testFailureWithPendingCoroutine() = runTest(dispatchTimeoutMs = 5000) {
    launch {
        withContext(NonCancellable) {
            awaitCancellation() // hanging non-cancellable coroutine
        }
    }
    yield()
    fail("some test failure") // this line is reached and throws, but it's never reported
}
```

I could see with debugging that the `fail()` is indeed reached and called. It does fail, but the assertion error is just swallowed somewhere in `runTest` and never makes it to the output. I just get the error message about the hanging coroutine:

```
The test coroutine was not completed after waiting for 5000 ms
kotlinx.coroutines.test.UncompletedCoroutinesError: The test coroutine was not completed after waiting for 5000 ms
	at app//kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt$runTestCoroutine$3$3.invokeSuspend(TestBuilders.kt:231)
	(Coroutine boundary)
	at kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt.runTestCoroutine(TestBuilders.kt:257)
	at kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt$runTest$1$1.invokeSuspend(TestBuilders.kt:167)
	at kotlinx.coroutines.test.TestBuildersJvmKt$createTestResult$1.invokeSuspend(TestBuildersJvm.kt:13)
Caused by: kotlinx.coroutines.test.UncompletedCoroutinesError: The test coroutine was not completed after waiting for 5000 ms
	at app//kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt$runTestCoroutine$3$3.invokeSuspend(TestBuilders.kt:231)
	at app//kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        ...
```

Moreover, the test spends the whole `dispatchTimeoutMs` before failing (this is why I reduced it to 5 seconds in the repro). It is somewhat understandable because of the non-cancellability of the coroutine, but it's quite annoying too. Not sure we could do anything about that though, at least the reporting of test failure in addition to the timeout would be great.

--- a/kotlinx-coroutines-core/common/src/internal/StackTraceRecovery.common.kt
+++ b/kotlinx-coroutines-core/common/src/internal/StackTraceRecovery.common.kt
@@ -40,6 +40,7 @@ internal expect suspend inline fun recoverAndThrow(exception: Throwable): Nothin
  * The opposite of [recoverStackTrace].
  * It is guaranteed that `unwrap(recoverStackTrace(e)) === e`
  */
+@PublishedApi
 internal expect fun <E: Throwable> unwrap(exception: E): E
 
 internal expect class StackTraceElement

--- a/kotlinx-coroutines-core/common/src/internal/StackTraceRecovery.common.kt
+++ b/kotlinx-coroutines-core/common/src/internal/StackTraceRecovery.common.kt
@@ -40,7 +40,7 @@ internal expect suspend inline fun recoverAndThrow(exception: Throwable): Nothin
  * The opposite of [recoverStackTrace].
  * It is guaranteed that `unwrap(recoverStackTrace(e)) === e`
  */
-@PublishedApi
+@PublishedApi // only published for the multiplatform tests in our own code
 internal expect fun <E: Throwable> unwrap(exception: E): E
 
 internal expect class StackTraceElement

--- a/kotlinx-coroutines-test/common/src/TestBuilders.kt
+++ b/kotlinx-coroutines-test/common/src/TestBuilders.kt
@@ -161,10 +161,11 @@ public fun runTest(
 public fun TestScope.runTest(
     dispatchTimeoutMs: Long = DEFAULT_DISPATCH_TIMEOUT_MS,
     testBody: suspend TestScope.() -> Unit
-): TestResult = asSpecificImplementation().let {
-    it.enter()
+): TestResult = with(asSpecificImplementation()) {
+    enter()
+    start(CoroutineStart.UNDISPATCHED, this, testBody)
     createTestResult {
-        runTestCoroutine(it, dispatchTimeoutMs, TestScopeImpl::tryGetCompletionCause, testBody) { it.leave() }
+        runTestCoroutine(this, dispatchTimeoutMs, { tryGetCompletionCause() }) { leave() }
     }
 }
 
@@ -190,22 +191,18 @@ internal const val DEFAULT_DISPATCH_TIMEOUT_MS = 60_000L
  * Run the [body][testBody] of the [test coroutine][coroutine], waiting for asynchronous completions for at most
  * [dispatchTimeoutMs] milliseconds, and performing the [cleanup] procedure at the end.
  *
+ * [tryGetCompletionCause] is the [JobSupport.completionCause], which is passed explicitly because it is protected.
+ *
  * The [cleanup] procedure may either throw [UncompletedCoroutinesError] to denote that child coroutines were leaked, or
  * return a list of uncaught exceptions that should be reported at the end of the test.
  */
-internal suspend fun <T: AbstractCoroutine<Unit>> runTestCoroutine(
-    coroutine: T,
+internal suspend fun runTestCoroutine(
+    coroutine: AbstractCoroutine<Unit>,
     dispatchTimeoutMs: Long,
-    tryGetCompletionCause: T.() -> Throwable?,
-    testBody: suspend T.() -> Unit,
+    tryGetCompletionCause: () -> Throwable?,
     cleanup: () -> List<Throwable>,
 ) {
     val scheduler = coroutine.coroutineContext[TestCoroutineScheduler]!!
-    /** TODO: moving this [AbstractCoroutine.start] call outside [createTestResult] fails on Native with
-     * [TestCoroutineDispatcher], because the event loop is not started. */
-    coroutine.start(CoroutineStart.UNDISPATCHED, coroutine) {
-        testBody()
-    }
     var completed = false
     while (!completed) {
         scheduler.advanceUntilIdle()
@@ -223,32 +220,7 @@ internal suspend fun <T: AbstractCoroutine<Unit>> runTestCoroutine(
                 // we received knowledge that `scheduler` observed a dispatch event, so we reset the timeout
             }
             onTimeout(dispatchTimeoutMs) {
-                try {
-                    cleanup()
-                } catch (e: UncompletedCoroutinesError) {
-                    // we expect these and will instead throw a more informative exception just below.
-                    emptyList()
-                }.throwAll()
-                var completing: Boolean
-                val completionCause = try {
-                    coroutine.tryGetCompletionCause().also { completing = true }
-                } catch (e: Throwable) {
-                    completing = false
-                    null
-                }
-                var message = "After waiting for $dispatchTimeoutMs ms"
-                if (!completing)
-                    message += ", the test coroutine is not completing"
-                val activeChildren = coroutine.children.filter { it.isActive }.toList()
-                if (activeChildren.isNotEmpty())
-                    message += ", there were active child jobs: $activeChildren"
-                if (completing && activeChildren.isEmpty()) {
-                    // some sort of race condition? write something generic.
-                    message += ", the test coroutine was not completed"
-                }
-                val error = UncompletedCoroutinesError(message)
-                completionCause?.let { cause -> error.addSuppressed(cause) }
-                throw error
+                handleTimeout(coroutine, dispatchTimeoutMs, tryGetCompletionCause, cleanup)
             }
         }
     }
@@ -262,6 +234,41 @@ internal suspend fun <T: AbstractCoroutine<Unit>> runTestCoroutine(
         (listOf(exception) + exceptions).throwAll()
     }
     cleanup().throwAll()
+}
+
+/**
+ * Invoked on timeout in [runTest]. Almost always just builds a nice [UncompletedCoroutinesError] and throws it.
+ * However, sometimes it detects that the coroutine completed, in which case it returns normally.
+ */
+private inline fun handleTimeout(
+    coroutine: AbstractCoroutine<Unit>,
+    dispatchTimeoutMs: Long,
+    tryGetCompletionCause: () -> Throwable?,
+    cleanup: () -> List<Throwable>,
+) {
+    val uncaughtExceptions = try {
+        cleanup()
+    } catch (e: UncompletedCoroutinesError) {
+        // we expect these and will instead throw a more informative exception.
+        emptyList()
+    }
+    val activeChildren = coroutine.children.filter { it.isActive }.toList()
+    val completionCause = if (coroutine.isCancelled) tryGetCompletionCause() else null
+    var message = "After waiting for $dispatchTimeoutMs ms"
+    if (completionCause == null)
+        message += ", the test coroutine is not completing"
+    if (activeChildren.isNotEmpty())
+        message += ", there were active child jobs: $activeChildren"
+    if (completionCause != null && activeChildren.isEmpty()) {
+        if (coroutine.isCompleted)
+            return
+        // TODO: can this really ever happen?
+        message += ", the test coroutine was not completed"
+    }
+    val error = UncompletedCoroutinesError(message)
+    completionCause?.let { cause -> error.addSuppressed(cause) }
+    uncaughtExceptions.forEach { error.addSuppressed(it) }
+    throw error
 }
 
 internal fun List<Throwable>.throwAll() {

--- a/kotlinx-coroutines-test/common/src/TestBuilders.kt
+++ b/kotlinx-coroutines-test/common/src/TestBuilders.kt
@@ -164,7 +164,7 @@ public fun TestScope.runTest(
 ): TestResult = asSpecificImplementation().let {
     it.enter()
     createTestResult {
-        runTestCoroutine(it, dispatchTimeoutMs, testBody) { it.leave() }
+        runTestCoroutine(it, dispatchTimeoutMs, TestScopeImpl::tryGetCompletionCause, testBody) { it.leave() }
     }
 }
 
@@ -196,6 +196,7 @@ internal const val DEFAULT_DISPATCH_TIMEOUT_MS = 60_000L
 internal suspend fun <T: AbstractCoroutine<Unit>> runTestCoroutine(
     coroutine: T,
     dispatchTimeoutMs: Long,
+    tryGetCompletionCause: T.() -> Throwable?,
     testBody: suspend T.() -> Unit,
     cleanup: () -> List<Throwable>,
 ) {
@@ -228,7 +229,26 @@ internal suspend fun <T: AbstractCoroutine<Unit>> runTestCoroutine(
                     // we expect these and will instead throw a more informative exception just below.
                     emptyList()
                 }.throwAll()
-                throw UncompletedCoroutinesError("The test coroutine was not completed after waiting for $dispatchTimeoutMs ms")
+                var completing: Boolean
+                val completionCause = try {
+                    coroutine.tryGetCompletionCause().also { completing = true }
+                } catch (e: Throwable) {
+                    completing = false
+                    null
+                }
+                var message = "After waiting for $dispatchTimeoutMs ms"
+                if (!completing)
+                    message += ", the test coroutine is not completing"
+                val activeChildren = coroutine.children.filter { it.isActive }.toList()
+                if (activeChildren.isNotEmpty())
+                    message += ", there were active child jobs: $activeChildren"
+                if (completing && activeChildren.isEmpty()) {
+                    // some sort of race condition? write something generic.
+                    message += ", the test coroutine was not completed"
+                }
+                val error = UncompletedCoroutinesError(message)
+                completionCause?.let { cause -> error.addSuppressed(cause) }
+                throw error
             }
         }
     }

--- a/kotlinx-coroutines-test/common/src/TestBuilders.kt
+++ b/kotlinx-coroutines-test/common/src/TestBuilders.kt
@@ -161,11 +161,10 @@ public fun runTest(
 public fun TestScope.runTest(
     dispatchTimeoutMs: Long = DEFAULT_DISPATCH_TIMEOUT_MS,
     testBody: suspend TestScope.() -> Unit
-): TestResult = with(asSpecificImplementation()) {
-    enter()
-    start(CoroutineStart.UNDISPATCHED, this, testBody)
+): TestResult = asSpecificImplementation().let {
+    it.enter()
     createTestResult {
-        runTestCoroutine(this, dispatchTimeoutMs, { tryGetCompletionCause() }) { leave() }
+        runTestCoroutine(it, dispatchTimeoutMs, TestScopeImpl::tryGetCompletionCause, testBody) { it.leave() }
     }
 }
 
@@ -196,13 +195,18 @@ internal const val DEFAULT_DISPATCH_TIMEOUT_MS = 60_000L
  * The [cleanup] procedure may either throw [UncompletedCoroutinesError] to denote that child coroutines were leaked, or
  * return a list of uncaught exceptions that should be reported at the end of the test.
  */
-internal suspend fun runTestCoroutine(
-    coroutine: AbstractCoroutine<Unit>,
+internal suspend fun <T: AbstractCoroutine<Unit>> runTestCoroutine(
+    coroutine: T,
     dispatchTimeoutMs: Long,
-    tryGetCompletionCause: () -> Throwable?,
+    tryGetCompletionCause: T.() -> Throwable?,
+    testBody: suspend T.() -> Unit,
     cleanup: () -> List<Throwable>,
 ) {
     val scheduler = coroutine.coroutineContext[TestCoroutineScheduler]!!
+    /** TODO: moving this [AbstractCoroutine.start] call outside [createTestResult] fails on JS. */
+    coroutine.start(CoroutineStart.UNDISPATCHED, coroutine) {
+        testBody()
+    }
     var completed = false
     while (!completed) {
         scheduler.advanceUntilIdle()
@@ -240,10 +244,10 @@ internal suspend fun runTestCoroutine(
  * Invoked on timeout in [runTest]. Almost always just builds a nice [UncompletedCoroutinesError] and throws it.
  * However, sometimes it detects that the coroutine completed, in which case it returns normally.
  */
-private inline fun handleTimeout(
-    coroutine: AbstractCoroutine<Unit>,
+private inline fun<T: AbstractCoroutine<Unit>> handleTimeout(
+    coroutine: T,
     dispatchTimeoutMs: Long,
-    tryGetCompletionCause: () -> Throwable?,
+    tryGetCompletionCause: T.() -> Throwable?,
     cleanup: () -> List<Throwable>,
 ) {
     val uncaughtExceptions = try {
@@ -253,7 +257,7 @@ private inline fun handleTimeout(
         emptyList()
     }
     val activeChildren = coroutine.children.filter { it.isActive }.toList()
-    val completionCause = if (coroutine.isCancelled) tryGetCompletionCause() else null
+    val completionCause = if (coroutine.isCancelled) coroutine.tryGetCompletionCause() else null
     var message = "After waiting for $dispatchTimeoutMs ms"
     if (completionCause == null)
         message += ", the test coroutine is not completing"

--- a/kotlinx-coroutines-test/common/src/TestScope.kt
+++ b/kotlinx-coroutines-test/common/src/TestScope.kt
@@ -216,6 +216,9 @@ internal class TestScopeImpl(context: CoroutineContext) :
         }
     }
 
+    /** Throws an exception if the coroutine is not completing. */
+    fun tryGetCompletionCause(): Throwable? = completionCause
+
     override fun toString(): String =
         "TestScope[" + (if (finished) "test ended" else if (entered) "test started" else "test not started") + "]"
 }

--- a/kotlinx-coroutines-test/common/test/RunTestTest.kt
+++ b/kotlinx-coroutines-test/common/test/RunTestTest.kt
@@ -5,6 +5,7 @@
 package kotlinx.coroutines.test
 
 import kotlinx.coroutines.*
+import kotlinx.coroutines.internal.*
 import kotlinx.coroutines.flow.*
 import kotlin.coroutines.*
 import kotlin.test.*
@@ -94,6 +95,59 @@ class RunTestTest {
                 3
             }
             fail("shouldn't be reached")
+        }
+    }
+
+    /** Tests that, on timeout, the names of the active coroutines are listed,
+     * whereas the names of the completed ones are not. */
+    @Test
+    @NoJs
+    @NoNative
+    fun testListingActiveCoroutinesOnTimeout(): TestResult {
+        val name1 = "GoodUniqueName"
+        val name2 = "BadUniqueName"
+        return testResultMap({
+            try {
+                it()
+                fail("unreached")
+            } catch (e: UncompletedCoroutinesError) {
+                assertTrue((e.message ?: "").contains(name1))
+                assertFalse((e.message ?: "").contains(name2))
+            }
+        }) {
+            runTest(dispatchTimeoutMs = 10) {
+                launch(CoroutineName(name1)) {
+                    CompletableDeferred<Unit>().await()
+                }
+                launch(CoroutineName(name2)) {
+                }
+            }
+        }
+    }
+
+    /** Tests that the [UncompletedCoroutinesError] suppresses an exception with which the coroutine is completing. */
+    @Test
+    fun testFailureWithPendingCoroutine() = testResultMap({
+        try {
+            it()
+            fail("unreached")
+        } catch (e: UncompletedCoroutinesError) {
+            @Suppress("INVISIBLE_MEMBER")
+            val suppressed = unwrap(e).suppressedExceptions
+            assertEquals(1, suppressed.size)
+            assertIs<TestException>(suppressed[0]).also {
+                assertEquals("A", it.message)
+            }
+        }
+    }) {
+        runTest(dispatchTimeoutMs = 10) {
+            launch {
+                withContext(NonCancellable) {
+                    awaitCancellation()
+                }
+            }
+            yield()
+            throw TestException("A")
         }
     }
 

--- a/kotlinx-coroutines-test/common/test/RunTestTest.kt
+++ b/kotlinx-coroutines-test/common/test/RunTestTest.kt
@@ -159,14 +159,24 @@ class RunTestTest {
         }
     }
 
-    /** Tests uncaught exceptions taking priority over dispatch timeout in error reports. */
+    /** Tests uncaught exceptions being suppressed by the dispatch timeout error. */
     @Test
     @NoNative // TODO: timeout leads to `Cannot execute task because event loop was shut down` on Native
     fun testRunTestTimingOutAndThrowing() = testResultMap({ fn ->
-        assertFailsWith<IllegalArgumentException> { fn() }
+        try {
+            fn()
+            fail("unreached")
+        } catch (e: UncompletedCoroutinesError) {
+            @Suppress("INVISIBLE_MEMBER")
+            val suppressed = unwrap(e).suppressedExceptions
+            assertEquals(1, suppressed.size)
+            assertIs<TestException>(suppressed[0]).also {
+                assertEquals("A", it.message)
+            }
+        }
     }) {
         runTest(dispatchTimeoutMs = 1) {
-            coroutineContext[CoroutineExceptionHandler]!!.handleException(coroutineContext, IllegalArgumentException())
+            coroutineContext[CoroutineExceptionHandler]!!.handleException(coroutineContext, TestException("A"))
             withContext(Dispatchers.Default) {
                 delay(10000)
                 3

--- a/kotlinx-coroutines-test/jvm/src/migration/TestBuildersDeprecated.kt
+++ b/kotlinx-coroutines-test/jvm/src/migration/TestBuildersDeprecated.kt
@@ -8,7 +8,6 @@
 package kotlinx.coroutines.test
 
 import kotlinx.coroutines.*
-import kotlinx.coroutines.selects.*
 import kotlin.coroutines.*
 import kotlin.jvm.*
 
@@ -137,9 +136,9 @@ public fun runTestWithLegacyScope(
 ): TestResult {
     if (context[RunningInRunTest] != null)
         throw IllegalStateException("Calls to `runTest` can't be nested. Please read the docs on `TestResult` for details.")
-    val testScope = TestBodyCoroutine<Unit>(createTestCoroutineScope(context + RunningInRunTest))
+    val testScope = TestBodyCoroutine(createTestCoroutineScope(context + RunningInRunTest))
     return createTestResult {
-        runTestCoroutine(testScope, dispatchTimeoutMs, testBody) {
+        runTestCoroutine(testScope, dispatchTimeoutMs, TestBodyCoroutine::tryGetCompletionCause, testBody) {
             try {
                 testScope.cleanup()
                 emptyList()
@@ -169,9 +168,9 @@ public fun TestCoroutineScope.runTest(
     block: suspend TestCoroutineScope.() -> Unit
 ): TestResult = runTestWithLegacyScope(coroutineContext, dispatchTimeoutMs, block)
 
-private class TestBodyCoroutine<T>(
+private class TestBodyCoroutine(
     private val testScope: TestCoroutineScope,
-) : AbstractCoroutine<T>(testScope.coroutineContext, initParentJob = true, active = true), TestCoroutineScope {
+) : AbstractCoroutine<Unit>(testScope.coroutineContext, initParentJob = true, active = true), TestCoroutineScope {
 
     override val testScheduler get() = testScope.testScheduler
 
@@ -187,4 +186,7 @@ private class TestBodyCoroutine<T>(
         )
 
     fun cleanup() = testScope.cleanupTestCoroutines()
+
+    /** Throws an exception if the coroutine is not completing. */
+    fun tryGetCompletionCause(): Throwable? = completionCause
 }

--- a/kotlinx-coroutines-test/jvm/src/migration/TestBuildersDeprecated.kt
+++ b/kotlinx-coroutines-test/jvm/src/migration/TestBuildersDeprecated.kt
@@ -137,9 +137,8 @@ public fun runTestWithLegacyScope(
     if (context[RunningInRunTest] != null)
         throw IllegalStateException("Calls to `runTest` can't be nested. Please read the docs on `TestResult` for details.")
     val testScope = TestBodyCoroutine(createTestCoroutineScope(context + RunningInRunTest))
-    testScope.start(CoroutineStart.UNDISPATCHED, testScope, testBody)
     return createTestResult {
-        runTestCoroutine(testScope, dispatchTimeoutMs, { testScope.tryGetCompletionCause() }) {
+        runTestCoroutine(testScope, dispatchTimeoutMs, TestBodyCoroutine::tryGetCompletionCause, testBody) {
             try {
                 testScope.cleanup()
                 emptyList()

--- a/kotlinx-coroutines-test/jvm/src/migration/TestBuildersDeprecated.kt
+++ b/kotlinx-coroutines-test/jvm/src/migration/TestBuildersDeprecated.kt
@@ -137,8 +137,9 @@ public fun runTestWithLegacyScope(
     if (context[RunningInRunTest] != null)
         throw IllegalStateException("Calls to `runTest` can't be nested. Please read the docs on `TestResult` for details.")
     val testScope = TestBodyCoroutine(createTestCoroutineScope(context + RunningInRunTest))
+    testScope.start(CoroutineStart.UNDISPATCHED, testScope, testBody)
     return createTestResult {
-        runTestCoroutine(testScope, dispatchTimeoutMs, TestBodyCoroutine::tryGetCompletionCause, testBody) {
+        runTestCoroutine(testScope, dispatchTimeoutMs, { testScope.tryGetCompletionCause() }) {
             try {
                 testScope.cleanup()
                 emptyList()

--- a/kotlinx-coroutines-test/jvm/test/migration/RunTestLegacyScopeTest.kt
+++ b/kotlinx-coroutines-test/jvm/test/migration/RunTestLegacyScopeTest.kt
@@ -6,6 +6,7 @@ package kotlinx.coroutines.test
 
 import kotlinx.coroutines.*
 import kotlinx.coroutines.flow.*
+import kotlinx.coroutines.internal.*
 import kotlin.coroutines.*
 import kotlin.test.*
 
@@ -100,10 +101,20 @@ class RunTestLegacyScopeTest {
 
     @Test
     fun testRunTestTimingOutAndThrowing() = testResultMap({ fn ->
-        assertFailsWith<IllegalArgumentException> { fn() }
+        try {
+            fn()
+            fail("unreached")
+        } catch (e: UncompletedCoroutinesError) {
+            @Suppress("INVISIBLE_MEMBER")
+            val suppressed = unwrap(e).suppressedExceptions
+            assertEquals(1, suppressed.size)
+            assertIs<TestException>(suppressed[0]).also {
+                assertEquals("A", it.message)
+            }
+        }
     }) {
         runTestWithLegacyScope(dispatchTimeoutMs = 1) {
-            coroutineContext[CoroutineExceptionHandler]!!.handleException(coroutineContext, IllegalArgumentException())
+            coroutineContext[CoroutineExceptionHandler]!!.handleException(coroutineContext, TestException("A"))
             withContext(Dispatchers.Default) {
                 delay(10000)
                 3

